modules/pam_limits/pam_limits.c

```diff
@@ -666,6 +702,19 @@ process_limit (const pam_handle_t *pamh, int source, const char *lim_type,
 	 rlimit_value = 20 - int_value;
          break;
 #endif
+	case RLIMIT_NOFILE:
+	/*
+	 * If nofile is to be set to "unlimited", try to set it to
+	 * the value in /proc/sys/fs/nr_open instead.
+	 */
+	if (rlimit_value == RLIM_INFINITY) {
+	    if (!value_from_proc_sys_fs("nr_open", &rlimit_value))
+		pam_syslog(pamh, LOG_WARNING,
+			   "Cannot set \"nofile\" to a sensible value");
+	    else if (ctrl & PAM_DEBUG_ARG)
+		pam_syslog(pamh, LOG_DEBUG, "Setting \"nofile\" limit to %lu", (long unsigned) rlimit_value);
+	}
```

Comment:
Starting from this point, the value assigned to `rlimit_value` by `value_from_proc_sys_fs` is used regardless of the return value of the latter, is it expected?

Reply:
Maybe it would make some sense to default to some reasonably large value? So failure of value_from_proc_sys_fs does not prevent login?

Reply:
The question is: which one?
No matter which one you choose, there will always be cases where the value which **should** be in `/proc/sys/fs/nr_open` is smaller. So I would really go for: "if we have no other value, take what the administrator has configured".

Reply:
But that value is RLIM_INFINITY which is unusable and will fail the setrlimit() call. And that effectively makes using infinity for RLIMIT_NOFILE unusable because you cannot depend on it being converted to reasonable value.

On the other hand value_from_proc_sys_fs() should not normally fail, so perhaps we could right away make its failure to mean failing the pam_limit pam_open_session().

Reply:
How about not change the value at all?
If the value is RLIM_INFINITY and value_from_proc_sys_fs() fails, set rlimit_value to pl->limits[limit_item].limit.rlim_cur or pl->limits[limit_item].limit.rlim_max, depending upon limit_type?

Reply:
Hmm... that would make some sense. @ldv-alt what's your opinion?

Reply:
If the contents of `/proc/sys/fs/nr_open` cannot be obtained, there is no easy way to find out the maximal value for `RLIMIT_NOFILE`. It could be probed, but the range supported by Linux kernel is too large to probe. How likely is the chance of failing to obtain the contents of `/proc/sys/fs/nr_open`?
- `/proc` is not mounted in the mount namespace: not likely, but possible;
- `/proc/sys/fs/nr_open` is available but fopen or fgets failed: unlikely, other serious problems are expected;
- `strtoul` failed: very unlikely, other serious problems are expected.

There are indeed two possible courses of action:
- strict: use the value of `RLIM_INFINITY` which is always larger than the maximal value for `RLIMIT_NOFILE`, causing the module to fail with `PAM_PERM_DENIED` as before the proposed change;
- permissive: avoid changing the corresponding limit as if it was not configured.

I'm slightly in favor of the first variant as it seems to follow the principle of the least surprise: signal an error if the maximal limit is configured but the software failed to calculate the corresponding value.